activation/activation_test.go

```diff
@@ -271,48 +159,87 @@ func assertLastAtx(r *require.Assertions, posAtx, prevAtx *types.VerifiedActivat
 	r.Equal(poetRef, atx.GetPoetProofRef())
 }
 
-func publishAtx(t *testing.T, b *Builder, clockEpoch types.EpochID, buildNIPostLayerDuration uint32) (published, builtNIPost bool, err error) {
+func publishAtx(
+	t *testing.T,
+	tab *testAtxBuilder,
+	posAtxId types.ATXID,
+	posAtxLayer types.LayerID,
+	currLayer *types.LayerID, // pointer to keep current layer consistent across calls
+	buildNIPostLayerDuration uint32,
+	publishErr error,
+) (*types.ActivationTx, error) {
 	t.Helper()
-	net.lastTransmission = nil
-
-	nipostBuilderMock.mu.Lock()
-	nipostBuilderMock.buildNIPostFunc = func(challenge *types.PoetChallenge) (*types.NIPost, time.Duration, error) {
-		builtNIPost = true
-		layerClockMock.mu.Lock()
-		layerClockMock.currentLayer = layerClockMock.currentLayer.Add(buildNIPostLayerDuration)
-		layerClockMock.mu.Unlock()
-		hash, err := challenge.Hash()
-		require.NoError(t, err)
-		return newNIPostWithChallenge(hash, poetBytes), 0, nil
-	}
-	nipostBuilderMock.mu.Unlock()
-
-	layerClockMock.mu.Lock()
-	layerClockMock.currentLayer = clockEpoch.FirstLayer().Add(3)
-	layerClockMock.mu.Unlock()
-
-	err = b.PublishActivationTx(context.Background())
 
-	nipostBuilderMock.mu.Lock()
-	nipostBuilderMock.buildNIPostFunc = nil
-	nipostBuilderMock.mu.Unlock()
-
-	return net.lastTransmission != nil, builtNIPost, err
+	publishEpoch := posAtxLayer.GetEpoch() + 1
+	tab.mclock.EXPECT().GetCurrentLayer().DoAndReturn(
+		func() types.LayerID {
+			return *currLayer
+		}).AnyTimes()
+	tab.mhdlr.EXPECT().GetPosAtxID().Return(posAtxId, nil)
+	tab.mclock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(
+		func(got types.LayerID) time.Time {
+			require.Equal(t, publishEpoch.FirstLayer(), got)
+			return time.Now().Add(100 * time.Second)
+		})
```

Comment:
Please, consider using the formula I came up with for #3924 :
```go
layerClock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(func(layer types.LayerID) time.Time {
		// time.Now() ~= currentLayer
		genesis := time.Now().Add(-time.Duration(currentLayer.Value) * layerDuration)
		return genesis.Add(layerDuration * time.Duration(layer.Value))
	}).AnyTimes()
```

It calculates a time for given layer based on the current layer and a `layerDuration` constant.

Reply:
i do it here and in other tests that publish atx. but for the wait for first atx test, i left the code as is.
thanks for the suggestion.